Documentation/design/renderphase.md

```diff
@@ -0,0 +1,129 @@
+
+### InstallConfig Validation
+
+Render needs to perform 3 kinds of validation:
+
+1. Validate configuration options are known. e.g. networking provider is known.
+
+2. Platform configuration validation. This also includes validating all node pools belong to the specified platform.
+
+3. Ensure `master` node pool exists and has valid configuration.
+
+### TLS asset generation
+
+Generate all the TLS assets required for the cluster and thier corresponding Kubernetes secrets. All this needs to be go code rather than terraform templates.
+
+Some of the TLS assets that need to be generated are:
+1. root-ca
+2. kube-ca
+3. aggregator-ca
+4. etcd-ca
+5. apiserver
+6. ingress and more
+
+### Rendering operator manifests
+
+1. Historically, the installer had manifests for all the cluster objects that are required to run the operator. For example, deployment manifest for the operator, service account, roles/rolebindings and operator config etc. So changes to the installer were required to test any change in the operator. This meant that the first indication if the operator had broken install was after the operator changes were merged in its repository and then operator was updated in the installer repository.
+
+2. Also, all the cluster objects required by the operator to functions had to be managed by the installer team.
+
+Therefore, to allow operator teams to iterate and control all aspects of an operator, the installer needs a process to delegate the `InstallConfig` object to each operator in render phase to generate iteself and all other cluster objects necessary for it to function correctly.
+
+#### Prerequesites
+
+1. Render cannot assume contanier runtime on host.
+2. Render needs to be successfull on atleast linux and darwin systems.
+
+#### Solutions
+
+1. Vendor operators
+
+    One way to achieve rendering is to vendor operator's code into installer and hook operator defined render function in installer's Render phase.
+
+    Pros:
+
+    * No extra work for cross-platfrom compile for various operators.
+    * Single simple go binary for installer.
+
+    Cons:
+
+    * On operator repository, CI needs to update the toml file and run re-vendor to run install test in their PR.
+
+2. Git submodules
+
+    Installer repo uses git submodules to import operators and then asserts **build and run interfaces** between installer and operator to hook operator into installer's Render phase.
+
+    Pros:
+
+    * Installer repo is smaller.
+    * CI is much easier to work as it doesn't have to deal with go vendoring and only git submodules to replace/update the operator in testing.
+
+    Cons:
+
+    * Build interface between installer and operator repo needs to account for cross-platform builds.
+    * Separate binaries for each operator will be generated, packing them into single binary **might** be challenging in go.
```

Comment:
Is "Render cannot assume network access on host." another requirement?  Otherwise "pack things into a single binary" seems like more trouble than it's worth.  You can have the installer fetch operators over the network at runtime (like [Terraform's providers][1]).  Or you could give the installer a tarball with builds for different platforms ([how the installer works now for the initial binary][2]?).

[1]: https://www.terraform.io/docs/extend/plugin-types.html#providers
[2]: https://github.com/openshift/installer/blob/ae41b0a3fa6f3c074bb1e0fd726a9aec406c4301/Documentation/dev/build.md#building-a-release-tarball

Reply:
There was some discussion of the "single binary" actually just being a self-extracting archive, so separate binaries is an option while still maintaining "simple" distribution. Not requiring network access for the render step would be nice - but I'm not sure something we've really run into (not often you want to generate the assets, but not use them - and using them more or less assumes network access).